// kissfft-wasm/lib/fftnd.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FFTNDConfig = void 0;
const types_1 = require("./types");
const wasm_1 = require("./wasm");
class FFTNDConfig extends types_1.KissFFTConfig {
    constructor(dims, inverse) {
        super(dims.reduce((x, y) => x * y), inverse);
        this.dims = dims;
        this.ptr = 0;
        const ndims = dims.length;
        const dimsPtr = wasm_1.wasm._malloc(Int32Array.BYTES_PER_ELEMENT * ndims);
        dims.forEach((x, i) => {
            wasm_1.wasm.HEAP32[dimsPtr / Int32Array.BYTES_PER_ELEMENT + i] = x;
        });
        this.ptr = wasm_1.wasm._kiss_fftnd_alloc(dimsPtr, ndims, inverse, 0, 0);
        wasm_1.wasm._free(dimsPtr);
    }
    get pointer() {
        return this.ptr;
    }
    free() {
        wasm_1.wasm._free(this.ptr);
        this.ptr = 0;
    }
    work(input, output) {
        this.check(input, output);
        wasm_1.wasm._kiss_fftnd(this.ptr, input.pointer, output.pointer);
        if (this.inverse) {
            wasm_1.wasm._scale(output.pointer, this.nfft * 2, 1.0 / this.nfft);
        }
    }
}
exports.FFTNDConfig = FFTNDConfig;
