// wasm-noise/wasm_noise.js
import * as wasm from './wasm_noise_bg.wasm';

let cachedTextDecoder = new TextDecoder('utf-8', { ignoreBOM: true, fatal: true });

cachedTextDecoder.decode();

let cachegetUint8Memory0 = null;
function getUint8Memory0() {
    if (cachegetUint8Memory0 === null || cachegetUint8Memory0.buffer !== wasm.memory.buffer) {
        cachegetUint8Memory0 = new Uint8Array(wasm.memory.buffer);
    }
    return cachegetUint8Memory0;
}

function getStringFromWasm0(ptr, len) {
    return cachedTextDecoder.decode(getUint8Memory0().subarray(ptr, ptr + len));
}
/**
*/
export class Perlin2D {

    static __wrap(ptr) {
        const obj = Object.create(Perlin2D.prototype);
        obj.ptr = ptr;

        return obj;
    }

    free() {
        const ptr = this.ptr;
        this.ptr = 0;

        wasm.__wbg_perlin2d_free(ptr);
    }
    /**
    * @param {number} x
    * @param {number} y
    * @returns {number}
    */
    get(x, y) {
        var ret = wasm.perlin2d_get(this.ptr, x, y);
        return ret;
    }
    /**
    * @param {number} x
    * @param {number} y
    * @param {number} z
    * @returns {number}
    */
    get3(x, y, z) {
        var ret = wasm.perlin2d_get3(this.ptr, x, y, z);
        return ret;
    }
    /**
    * @returns {Perlin2D}
    */
    static new() {
        var ret = wasm.perlin2d_new();
        return Perlin2D.__wrap(ret);
    }
}

export const __wbindgen_throw = function(arg0, arg1) {
    throw new Error(getStringFromWasm0(arg0, arg1));
};
